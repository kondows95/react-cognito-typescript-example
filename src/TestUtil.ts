import {ReactElement} from 'react';
import { render, fireEvent, RenderResult } from '@testing-library/react';
export default class TestUtil
{
  public render: RenderResult;
  constructor(component:ReactElement) {
    this.render = render(component);
  }

  setValue(testId: string, value: string | number) {
    const elem = this.render.getByTestId(testId);
    fireEvent.change(elem, {target: {value: value}})
  }

  click(testId: string) {
    const elem = this.render.getByTestId(testId);
    fireEvent.click(elem);
  }

  getAll(testId: string) {
    return this.render.getAllByTestId(testId);
  }

  get(testId: string) {
    return this.render.getByTestId(testId);
  }
}