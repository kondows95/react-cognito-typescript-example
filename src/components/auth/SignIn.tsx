import React, {MouseEvent, FormEvent, ChangeEvent} from 'react';
import { Box, Container, Link, TextField}  from '@material-ui/core';
import FormTitle from './FormTitle';
import SubmitButton from './SubmitButton';

type Props = {
  authState: string
  changeAuthState: (state: string) => void
  signIn: (email: string, password: string) => void
  loading: boolean
  error: string
};

const SignIn: React.FC<Props> = ({ authState, changeAuthState, signIn, loading, error }) => {
  type Form = {　[key: string]: string　};
  const [form, setForm] = React.useState<Form>({email:"", password:""});
  
  type FormKey = "email" | "password";
  const handleChangeValue = (fieldName: FormKey) => (event: ChangeEvent<HTMLInputElement>) => {
    const newForm = {...form};
    newForm[fieldName] =  event.currentTarget.value;
    setForm(newForm);
  };
  
  const handleSubmit = (event: FormEvent)  => {
    event.preventDefault();
    signIn(form['email'], form['password']);
  };

  const handleForgotPassword = (event: MouseEvent)  => {
    event.preventDefault();
    changeAuthState('forgotPassword');
  };

  const content = (
    <Container component="main" maxWidth="xs">
      <form onSubmit={handleSubmit}>
        <Box display="flex" flexDirection="column" mt={8}>
          <FormTitle>Sign In</FormTitle>
          <Box display="flex" justifyContent="center" fontWeight={600} color="error.main">
            {error}
          </Box>
          <Box width="100%" my={2}>
            <TextField
              autoComplete="email"
              type="email"
              onChange={handleChangeValue('email')}
              value={form.email}
              label={loading? " " : "Email Address"}
              variant="outlined"
              required
              fullWidth
              inputProps={{"data-testid":"email"}}
            />
          </Box>
          <Box width="100%" my={2}>
            <TextField
              label={loading? " " : "Password"}
              type="password"
              autoComplete="current-password"
              onChange={handleChangeValue('password')} 
              value={form.password}
              variant="outlined"
              required
              fullWidth
              inputProps={{"data-testid":"password"}}
            />
          </Box>
          <Box width="100%" mt={4} mb={2} className="relative">
            <SubmitButton　data-testid="submitButton" disabled={loading}>
              Sign In
            </SubmitButton>
          </Box>
          <Box width="100%" my={2}>
            <Link data-testid="forgotPasswordLink" href="#" variant="body2" onClick={handleForgotPassword}>
              Forgot password?
            </Link>
          </Box>
        </Box>
      </form>
    </Container>
  );
  return (authState === 'signIn') ? content : null;
};

export default SignIn;