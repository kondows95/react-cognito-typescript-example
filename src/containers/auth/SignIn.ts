import { connect } from 'react-redux';
import SignIn from '../../components/auth/SignIn';
import { AppState, AppDispatch } from '../../modules/index';
import { changeAuthState, signIn } from '../../modules/auth';

export const mapStateToProps = (state: AppState) => ({
  authState: state.auth.authState,
  loading: state.auth.loading,
  error: state.auth.error,
});

export const mapDispatchToProps = (dispatch: AppDispatch) => ({
  changeAuthState: (value: string) => dispatch(changeAuthState(value)),
  signIn: (email: string, password: string) => dispatch(signIn(email, password))
});

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);


