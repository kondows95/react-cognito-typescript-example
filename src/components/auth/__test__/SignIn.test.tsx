import React from 'react';
import SignIn from '../SignIn';
import TestUtil from '../../../TestUtil'; 

let util: TestUtil;
let changeAuthState: jest.Mock;
let signIn: jest.Mock;
beforeEach(() => {
  changeAuthState = jest.fn();
  signIn = jest.fn();
});

describe("SignIn Component",()=>{
  beforeEach(() => {
    util= new TestUtil(
      <SignIn
        authState={"signIn"}
        changeAuthState={changeAuthState}
        signIn={signIn}
        loading={false}
        error={""}
      />
    );
  });

  it('Snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('Click forgotPassword link', () => {
    util.click('forgotPasswordLink');
    expect(changeAuthState).toHaveBeenCalled();
  });

  it('Click submit button', () => {
    util.setValue('email', 'test@example.com');
    util.setValue('password', 'myPassword');
    util.click('submitButton');
    expect(signIn).toHaveBeenCalledWith("test@example.com","myPassword");
  });
});

describe("SignIn Component that loading is true",()=>{
  beforeEach(() => {
    util= new TestUtil(
      <SignIn
        authState={"forgotPassword"}
        changeAuthState={changeAuthState}
        signIn={signIn}
        loading={true}
        error={""}
      />
    );
  });

  it('Snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });
});
