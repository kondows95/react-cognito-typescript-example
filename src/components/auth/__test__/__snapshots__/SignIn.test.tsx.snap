// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Component Snapshot 1`] = `
<DocumentFragment>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <form>
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <div
          class="MuiBox-root MuiBox-root-10"
        >
          <div
            class="MuiBox-root MuiBox-root-11"
          >
            <div
              class="MuiBox-root MuiBox-root-12"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <g
                  fill="none"
                >
                  <path
                    d="M0 0h24v24H0V0z"
                  />
                  <path
                    d="M0 0h24v24H0V0z"
                    opacity=".87"
                  />
                </g>
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-22"
          >
            Sign In
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-23"
        />
        <div
          class="MuiBox-root MuiBox-root-24"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
            >
              Email Address
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="email"
                class="MuiInputBase-input MuiOutlinedInput-input"
                data-testid="email"
                required=""
                type="email"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-87 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-88"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-89"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
            >
              Password
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input"
                data-testid="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-87 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-88"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-90 relative"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
            data-testid="submitButton"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Sign In
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root MuiBox-root-124"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary"
            data-testid="forgotPasswordLink"
            href="#"
          >
            Forgot password?
          </a>
        </div>
      </div>
    </form>
  </main>
</DocumentFragment>
`;

exports[`SignIn Component that loading is true Snapshot 1`] = `<DocumentFragment />`;
